### ISSUE TITLE
Option to disallow unused checker suppressions

### ISSUE THREAD

**Issue by ben-manes** [google/error-prone#886]
As code evolves, some suppressions may be added that become unnecessary later one. However, these are easy to miss during refactoring and remains as benign. Unfortunately, this leaves open the possibility of a new error being introduced and uncaught. It can be unfeasible to audit a large code base for unnecessary suppressions after the fact, and generally shouldn't be an error as each tool has its own subset.

It would be nice if ErrorProne allowed its checks to be verified when suppressed. If the suppression is enabled but no error was reported, then the failure it reported. This would be hooked in [ErrorProneScanner](https://github.com/google/error-prone/blob/master/check_api/src/main/java/com/google/errorprone/scanner/ErrorProneScanner.java#L489) and disabled by default.

A scenario is null data flow analysis (https://github.com/uber/NullAway/issues/100) where false positives are suppressed. Then as the code or checker evolves these become potential hazards.

**Comment by jbduncan** [comment]
I don't really have anything to add to this other than I think this is, at least in theory, a great idea! I've no idea how practical it would be, but considering just the initial idea here, I'd be very, very happy to see something like this implemented.

**Comment by cushon** [comment]
This is something we've discussed, and I agree it would be nice to have. The approach I'd take would be similar to the one described here: https://github.com/typetools/checker-framework/issues/1701#issuecomment-349403873

**Comment by msridhar** [comment]
@cushon the naive approach I was thinking about was to just tweak the `ErrorProneScanner` code to look something like (for `AssignmentTree`s):
```java
    for (AssignmentTreeMatcher matcher : assignmentMatchers) {
      if (!isSuppressed(matcher, state)) {
        try {
          reportMatch(matcher.matchAssignment(tree, state), tree, state);
        } catch (Throwable t) {
          handleError(matcher, t);
        }
      } else if (checkSuppressionValidity) {
        if (!Description.NO_MATCH.equals(matcher.matchAssignment(tree, state))) {
          // warn about unnecessary suppression
        }
      }
    }
```
I don't think this would hurt performance much when not checking for valid suppressions.  It does clutter the ErrorProneScanner code a bit but I think it's still readable (the `else if` logic could be extracted to a function).  I might be missing something here, though.
